### PR TITLE
set up integration test framework

### PR DESCRIPTION
**Pull request by ryanfkeepers**
## Description

Adds the base test framework for handling
incremental backups integration testing.  Future
changes will add more tests to the table.

## Does this PR need a docs update or release note?

- [x] :no_entry: No 

## Type of change

- [x] :robot: Test

## Issue(s)

* #1966

## Test Plan

- [x] :green_heart: E2E

### CODE REVIEW

File: src/internal/operations/backup_integration_test.go [hunk 1]
```diff
@@ -191,67 +206,210 @@ func checkMetadataFilesExist(
...
+//revive:disable:context-as-argument
+func generateContainerOfItems(
+	t *testing.T,
+	ctx context.Context,
+	gc *connector.GraphConnector,
+	service path.ServiceType,
+	cat path.CategoryType,
+	sel selectors.Selector,
+	tenantID, userID, destFldr string,
+	howMany int,
```
Comment (dadams39) [inline src/internal/operations/backup_integration_test.go:296]:
```suggestion
     total int,
```
The variable name would need to be updated within the function.

File: src/internal/operations/backup_integration_test.go [hunk 2]
```diff
@@ -191,67 +206,210 @@ func checkMetadataFilesExist(
...
-	assert.NoError(t, err)
+	collections := []incrementalCollection{{
+		pathElements: []string{destFldr},
+		category:     cat,
+		items:        items,
+	}}
 
-	for _, col := range cols {
-		itemNames := []string{}
+	dest := control.DefaultRestoreDestination(common.SimpleTimeTesting)
```
Comment (dadams39) [inline src/internal/operations/backup_integration_test.go:325]: Weird thought, we call this almost the same through out the code. 
In a different PR should we just make it into the following 2:
`control.DefaultRestoreTestingDestination()`
`control.DefaultRestoreDestination()` 
~jat

File: src/internal/operations/backup_integration_test.go [hunk 3]
```diff
@@ -191,67 +206,210 @@ func checkMetadataFilesExist(
...
+	dest.ContainerName = destFldr
 
-		for item := range col.Items() {
-			assert.Implements(t, (*data.StreamSize)(nil), item)
```
Comment (dadams39) [inline src/internal/operations/backup_integration_test.go:232]: Is the test losing anything by removing this assertion? 

L334 has an error return, is that the error descriptive enough to lose the assertion?

File: src/internal/operations/backup_integration_test.go [hunk 4]
```diff
@@ -191,67 +206,210 @@ func checkMetadataFilesExist(
...
-			)
+	return gc.RestoreDataCollections(ctx, sel, dest, dataColls)
+}
 
-			itemNames = append(itemNames, item.UUID())
-		}
+type incrementalItem struct {
+	name string
+	data []byte
+}
```
Comment (dadams39) [inline src/internal/operations/backup_integration_test.go:344]: I could see the utility of this being in a global package, maybe even mocks or something. It works similarly to the `data.Stream`.  

Not necessary to fix. Just highlighting that within testing we continue to have to make these structs and if they just happened to be in the tool belt, well, they would be available for us within testing.

File: src/internal/operations/backup_integration_test.go [hunk 5]
```diff
@@ -191,67 +206,210 @@ func checkMetadataFilesExist(
...
 
-		assert.ElementsMatchf(
+type incrementalCollection struct {
+	// Elements (in order) for the path representing this collection. Should
+	// only contain elements after the prefix that corso uses for the path. For
+	// example, a collection for the Inbox folder in exchange mail would just be
+	// "Inbox".
```
Comment (dadams39) [inline src/internal/operations/backup_integration_test.go:350]: isn't this just the `path.folders()` call? 

https://github.com/alcionai/corso/blob/20ec708ea357c26afe0076e31f05eacb8597a688/src/pkg/path/resource_path.go#L206

File: src/internal/operations/backup_integration_test.go [hunk 6]
```diff
@@ -191,67 +206,210 @@ func checkMetadataFilesExist(
...
+}
+
+func toDataLayerPath(
+	t *testing.T,
+	service path.ServiceType,
+	tenant, user string,
+	category path.CategoryType,
+	elements []string,
+	isItem bool,
+) (path.Path, error) {
```
Comment (dadams39) [inline src/internal/operations/backup_integration_test.go:400]: Not sure how much we are leveraging the `testing.T` item, but we could lose the error return if `require` was used. Additionally, there are several `toDataLayer`-type names within path, would it be beneficial to refactor this to `fetchTestPath` or something of the like?

File: src/internal/operations/backup_integration_test.go [hunk 7]
```diff
@@ -451,6 +611,165 @@ func (suite *BackupOpIntegrationSuite) TestBackup_Run_exchange() {
...
+	)
+
+	m365, err := acct.M365Config()
+	require.NoError(t, err)
+
+	gc, err := connector.NewGraphConnector(ctx, acct, connector.Users)
+	require.NoError(t, err)
+
+	// generate 2 new folders and add data to them.
+	// need 2 points of data in each folder for testing purposes.
```
Comment (dadams39) [inline src/internal/operations/backup_integration_test.go:642]:
```suggestion
	// require 2 folders with identical data for testing purposes.
```

File: src/internal/operations/backup_integration_test.go [hunk 8]
```diff
@@ -451,6 +611,165 @@ func (suite *BackupOpIntegrationSuite) TestBackup_Run_exchange() {
...
+		},
+	}
+	for _, test := range table {
+		suite.T().Run(test.name, func(t *testing.T) {
+			var (
+				incMB = evmock.NewBus()
+				incBO = newBackupOp(t, ctx, kw, ms, acct, sel.Selector, incMB, ffs, closer)
+			)
+
+			test.updateUserData(t)
```
Comment (dadams39) [inline src/internal/operations/backup_integration_test.go:739]: Is this necessary for this iteration as it is empty?

File: src/internal/tester/cli.go [hunk 9]
```diff
@@ -36,7 +36,6 @@ func StubRootCmd(args ...string) *cobra.Command {
 func NewContext() (context.Context, func()) {
 	level := logger.Info
 
-	fmt.Printf("\n-----\nargs %v\n-----\n", os.Args)
```
Comment (dadams39) [inline src/internal/tester/cli.go:39]: From `src/internal/events/mock/mock.go`
+1 on the check annotations:
```go
type MockBus struct {
```
```bash
exported: type name will be used as mock.MockBus by other packages, and that stutters; consider calling this Bus (revive)
```

**Review by dadams39 — APPROVED** [review]
These are all nits and can be fixed here or in clean-up PRs. 
Overall, there are a few variable names that could be more descriptive and a few functions/structs that seem very similar to an existing framework buried in the code base. If the code is pushed without the changes, be sure to add a few issues for `tech-debt`.

File: src/internal/operations/backup_integration_test.go (hunk 2 above)
Comment (ryanfkeepers) [reply src/internal/operations/backup_integration_test.go:325]: It's bad form to add testing funcs to a standard package.  With how little we need to produce and modify here, I don't think there's anything wrong with using the standard constructor.

File: src/internal/operations/backup_integration_test.go (hunk 4 above)
Comment (ryanfkeepers) [reply src/internal/operations/backup_integration_test.go:344]: Generally agreed.  Although, and I haven't reviewed all the variations recently, it's possible that each instance is configured to fulfill some specific needs.  That would provide sufficient utility to maintain the apparent duplication.

File: src/internal/operations/backup_integration_test.go (hunk 6 above)
Comment (ryanfkeepers) [reply src/internal/operations/backup_integration_test.go:400]: Per error vs assert/require: sure, that sounds good.

Per the `toDataLayer`, I've talked with ash about a QoL constructor in paths.  The idea has grounds, but isn't of much importance.  Might make a tech debt issue for it later.

File: src/internal/operations/backup_integration_test.go (hunk 8 above)
Comment (ryanfkeepers) [reply src/internal/operations/backup_integration_test.go:739]: It'll be non-empty in all subsequent iterations.

File: src/internal/tester/cli.go (hunk 9 above)
Comment (ryanfkeepers) [reply src/internal/tester/cli.go:39]: good old stutters.

File: src/internal/operations/backup_integration_test.go (hunk 3 above)
Comment (ryanfkeepers) [reply src/internal/operations/backup_integration_test.go:232]: We didn't lose this assertion.  The diff is a bit weird, but the check is still there.

**Pull request merged**